### PR TITLE
Improve GNSS Device

### PR DESCRIPTION
**Pull request by Ellerbach**
## Description

Improve GNSS Device

## Motivation and Context

- adding more NMEA elements
- adding more tests
- correcting bugs
- changing visibility for NMEA parser to allow reuse
- adding new UnitsNet speed

## How Has This Been Tested?

Tests

## Types of changes

- [*] Improvement (non-breaking change that improves a feature, code or algorithm)
- [*] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [*] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [*] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [*] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [*] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [*] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

File: devices/GnssDevice/GprmcData.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+{
+    /// <summary>
+    /// Represents the GPRMC NMEA0183 data from a Gnss device. 
+    /// </summary>
+    public class GprmcData : INmeaData
+    {
+        /// <inheritdoc/>
+        public INmeaData Parse(string inputData)
+        {
+            var subfields = inputData.Split(',');
```
Comment (torbacz) [inline devices/GnssDevice/GprmcData.cs:18]: Looks like this one is the only one that throws exception. I think every data should behave the same, so either we add exceptions to all or we delete it from there.
I would rather go with exceptions. Let me know what you think and we can adjust code in next PR (I can do it)

File: devices/GnssDevice/Tests/GnssDeviceTests/NMEA0183ParserTests.cs [hunk 2]
```diff
@@ -26,7 +26,7 @@ public void ParseGngsa(string command, byte expectedMode, byte expectedFix)
         }
 
         [TestMethod]
-        [DataRow("$GPGLL,5109.0262317,N,11401.8407304,W,202725.00,A,D*79", 51.1504372f, -114.03067884f)]
```
Comment (torbacz) [inline devices/GnssDevice/Tests/GnssDeviceTests/NMEA0183ParserTests.cs:29]: Maybe add new DataRow instead of editing existing?

File: devices/GnssDevice/Tests/GnssDeviceTests/NMEA0183ParserTests.cs [hunk 3]
```diff
@@ -38,7 +38,7 @@ public void ParseGpgll(string command, float expectedLatitude, float expectedLon
         }
 
         [TestMethod]
-        [DataRow("$GPGGA,002153.000,3342.6618,N,11751.3858,W,1,10,1.2,27.0,M,-34.2,M,,0000*5E", 33.7110291f, -23.5230961f, 27.0f, 1.2f, 1313000d)]
```
Comment (torbacz) [inline devices/GnssDevice/Tests/GnssDeviceTests/NMEA0183ParserTests.cs:41]: Maybe add new DataRow instead of editing existing?

**Review by torbacz — COMMENTED** [review]
Just few tips :)

File: devices/GnssDevice/Tests/GnssDeviceTests/NMEA0183ParserTests.cs (hunk 2 above)
Comment (Ellerbach) [reply devices/GnssDevice/Tests/GnssDeviceTests/NMEA0183ParserTests.cs:29]: if you have real ones, give them to me :-) Happy to add more! . Same for next one.
I just bought a module to run more tests as well.

File: devices/GnssDevice/GprmcData.cs (hunk 1 above)
Comment (Ellerbach) [reply devices/GnssDevice/GprmcData.cs:18]: yea, I started to add this as it can be used outside of the parser.
So, I can add everywhere.
And I kind of like the idea of adding like initially in the interface the message name. As again, this can be used directly and you can know what it is parsing. What do you think?

File: devices/GnssDevice/Tests/GnssDeviceTests/NMEA0183ParserTests.cs (hunk 2 above)
Comment (torbacz) [reply devices/GnssDevice/Tests/GnssDeviceTests/NMEA0183ParserTests.cs:29]: Hmm, I thought It was real from my tests... Nevermind then, If anythink breaks durring my tests of real GPS device, I'll add more tests.

File: devices/GnssDevice/GprmcData.cs (hunk 1 above)
Comment (torbacz) [reply devices/GnssDevice/GprmcData.cs]: Anything that can improve current implementation will be great. 
For example, check this one:
https://github.com/nanoframework/nanoFramework.IoT.Device/pull/1088/files#diff-32766b81852d3246aa0b12375d048c0c331cdcbc237bf29b9a1a781fefe8044eR104

But maybe better idea is to add static field in each class - just to reference "ID" of the data?

**Comment by Ellerbach** [comment]
@torbacz reading more about all this, GGA, GSA, etc, etc standard messages and the prefix GP, BD, GN, and few more can change. So, We can even make all this more generic and the prefix used will show what system is used.
I will think of it and most likely also implement this in this PR.
 * GP = GPS
* GL = GLONASS
* GA = Galileo
* BD = BeiDou
* GI = NavIC
* GQ = QZSS
* GN = multiple systems (what is in fact formally called GNSS)

**Comment by Ellerbach** [comment]
@torbacz I finally created an abstract class, added couple of function there, extracted the message ID. So quite some rework. All seems ok by now and it's going to work for any positioning system!
The todo is to add few more tests, espcially broken positioning, things like this.
And extract the mode (GPS, GNSS, etc,) based on the 2 first letters. First, I need your feedback on the work here!

**Review by torbacz — COMMENTED** [review]
All good!

**Comment by Ellerbach** [comment]
@torbacz I've been very quickly added a `GenericSerialGnssDevice`, it should perfectly work for your use case (I had no time to test it), so please go ahead and test. It should just cover 95% of the scenarions. Still few details to add like the GnssMode (which has to be added into the recevie event as well with the same logic as the rest)

File: devices/GnssDevice/GprmcData.cs (hunk 1 above)
Comment (Ellerbach) [reply devices/GnssDevice/GprmcData.cs]: I've internalised in the GnssDevice and in a GenericSerialGnssDevice all the logic.

**Comment by Ellerbach** [comment]
@torbacz I made couple of additional changes and I'm fixing some of the edge cases. I still need to add more unit tests and fix more conversion elements.
I've been testing very quickly on a real device, and that's why I managed to fix things :-)

**Comment by Ellerbach** [comment]
@torbacz I've now tested with a real module, made some adjustments, added the doc, a full sample, including with extensibility for another type of message.
I guess now, it's about using it and seeing how it behaves.
I would still like to add:
*  few Unit Tests.
* an event to send the unparsed messages (similar to the parsed one)
* possibly add the validation of a message with checksum to avoid half formed messages (happened to me few times)

**Comment by Ellerbach** [comment]
OK, so UnitTests are enough in fact, seems they catch the issues when adding some little core. I've added a validation and an event.
So, now, it's fully ready for review :-)

**Comment by torbacz** [comment]
Tested with ATGM336H device, basic flow works fine. I saw one exception from serial (CLR_E_WRONG_TYPE) but only once. Then works great. 
Few commands are missing (GSV,ZDA,TXT), but it can be done in a future. 

```
Position: 49.xxxxxxx,20.xxxxxxxxxxxxxxx
No parser found for: $GPGSV,3,1,11,02,58,162,29,03,73,300,35,04,40,207,37,06,05,313,,0*67
No parser found for: $GPGSV,3,2,11,09,07,220,,17,36,284,27,19,25,311,28,21,41,153,33,0*65
No parser found for: $GPGSV,3,3,11,28,29,066,35,31,30,106,34,32,08,047,37,0*5D
No parser found for: $BDGSV,3,1,10,06,,,30,09,40,087,26,11,58,221,29,14,,,32,0*77
No parser found for: $BDGSV,3,2,10,21,60,059,21,28,,,27,36,,,33,39,,,27,0*4B
No parser found for: $BDGSV,3,3,10,42,64,251,36,43,,,34,0*42
No parser found for: $GNZDA,105708.000,11,08,2024,00,00*4F
No parser found for: $GPTXT,01,01,01,ANTENNA OK*35
````

Some minor changes from me (submitedPR against this branch).

**Comment by torbacz** [comment]
https://github.com/nanoframework/nanoFramework.IoT.Device/pull/1114

**Review by torbacz — APPROVED** [review]
Wow! That's a lot of code. Great job!

**Comment by Ellerbach** [comment]
> Few commands are missing (GSV,ZDA,TXT), but it can be done in a future

Yes, the GSV is not critical, it's all the satellites. I've added TXT in the sample. So anyone who needs can add.
For ZDA, we can add. It's an easy one!

File: devices/GnssDevice/GenericSerialGnssDevice.cs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            _serialPort = serialPort;
+            _shouldDispose = shouldDispose;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="GenericSerialGnssDevice"/> class.
+        /// </summary>
+        /// <param name="portName">Serial port name.</param>
+        /// <param name="baudRate">Baud rate, default to 9600.</param>
+        /// <param name="parity">PArity, default to None.</param>
```
Comment (josesimoes) [inline devices/GnssDevice/GenericSerialGnssDevice.cs:36]:
```suggestion
        /// <param name="parity">Parity, default to <see cref="Parity.None"/>.</param>
```

File: devices/GnssDevice/GnssDevice.cs [hunk 5]
```diff
@@ -77,29 +127,219 @@ protected set
...
+                }
+
                 _location = value;
                 LocationChanged?.Invoke(_location);
             }
         }
 
         /// <summary>
-        /// Represents the event handler for when the fix status of the Gnss module changes.
+        /// Gets or sets the satellites in view.
```
Comment (josesimoes) [inline devices/GnssDevice/GnssDevice.cs:146]: The description seems to differ from the property name... should be either in view or in use... don't know the details, so not sure which one is correct.

File: devices/GnssDevice/README.md [hunk 6]
```diff
@@ -1,3 +1,203 @@
-# Iot.Device.GnsssDevice
+# Global Navigation Satellite System Device NMEA 0183 - Including Generic Serial Module with GPS, GNSS, BeiDou
 
-Base Gnss classes for all NF Gnss modules. See [Atgm336h](../Atgm336h/) for a detailed usage.
+This bonding implement a base Global Navigation Satellite System (GNSS) device and a generic serial implementation. The device includes a generic serial module that supports multiple satellite navigation systems such as GPS, GNSS, and BeiDou. The Generic Serial GNSS Device is also extensible.
```
Comment (josesimoes) [inline devices/GnssDevice/README.md:3]:
```suggestion
This binding implements a base Global Navigation Satellite System (GNSS) device and a generic serial implementation. The device includes a generic serial module that supports multiple satellite navigation systems such as GPS, GNSS, and BeiDou. The Generic Serial GNSS Device is also extensible.
```

**Review by josesimoes — COMMENTED** [review]
Pretty nice implementation! 
I can see the ease to expand the parsers becoming very handy for complex use cases.
Great work guys.

Minor tiny bits and we're good to go. 💯
